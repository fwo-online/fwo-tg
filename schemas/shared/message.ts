import type { CharacterPublic } from '@/character/characterPublic';
import type { Character } from '@/character';
import type { Player } from './player';
import type { GameStatus } from '@/game';
import type { Action } from './action';
import type { RPC } from './rpc';

export type ClientToServerMessage = Message<{
  character: [callback: (character: Character) => void];
  'lobby:enter': [callback: (characters: CharacterPublic[]) => void];
  'lobby:leave': [];
  'lobby:start': [];
  'lobby:stop': [];
  'game:connected': [callback: (payload: RPC<{ players: Record<string, Player> }>) => void];
  'game:order': [
    order: {
      power: number;
      action: string;
      target: string;
    },
    callback: (
      payload: RPC<{
        actions: Action[];
        magics: Action[];
        skills: Action[];
        power: number;
        orders: {
          power: number;
          action: string;
          target: string;
        }[];
      }>,
    ) => void,
  ];
}>;

type Message<T extends Record<string, unknown[]>> = {
  [K in keyof T]: (...args: T[K]) => void;
};

export type ServerToClientMessage = Message<{
  character: [character: Character];
  'lobby:enter': [character: CharacterPublic];
  'lobby:leave': [character: CharacterPublic];
  'lobby:list': [characters: CharacterPublic[]];
  'lobby:start': [character: CharacterPublic];
  'lobby:stop': [character: CharacterPublic];
  'game:start': [gameID: string];
  'game:end': [];
  'game:startOrders': [actions: { actions: Action[]; magics: Action[]; skills: Action[] }];
  'game:endOrders': [];
  'game:startRound': [
    {
      round: number;
      status: Partial<Record<string, GameStatus[]>>;
    },
  ];
  'game:endRound': [{ dead: Player[] }];
  'game:kick': [{ reason: string; player: Player }];
  'game:preKick': [{ reason: string; player: Player }];
}>;
