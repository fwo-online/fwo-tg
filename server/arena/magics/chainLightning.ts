import { floatNumber } from '@/utils/floatNumber';
import { AoeDmgMagic } from '../Constuructors/AoeDmgMagicConstructor';
import type GameService from '../GameService';
import type { Player } from '../PlayersService';

/**
 * Цепь молний
 * Основное описание магии общее требовани есть в конструкторе
 */
class ChainLightning extends AoeDmgMagic {
  constructor() {
    super({
      name: 'chainLightning',
      displayName: 'Цепь молний',
      desc: 'Цепная молния повреждает выбранную цель молнией и еще одну случайно.',
      cost: 8,
      baseExp: 12,
      costType: 'mp',
      lvl: 3,
      orderType: 'any',
      aoeType: 'targetAoe',
      magType: 'bad',
      chance: [92, 94, 95],
      effect: ['1d3+1', '1d3+2', '1d3+3'],
      dmgType: 'lighting',
      profList: ['m'],
    });
  }

  maxTargets = [3, 4, 5];

  getTargets() {
    const { initiator, target, game } = this.params;
    const magicLevel = initiator.getMagicLevel(this.name);
    const maxTargets = this.maxTargets[magicLevel - 1];

    return game.players
      .getPlayersByClan(target.clan?.id)
      .filter(({ alive }) => alive)
      .filter(({ id }) => id !== target.id)
      .slice(0, maxTargets - 1);
  }

  /**
   * Основная функция запуска магии
   */
  run(initiator: Player, target: Player, game: GameService): void {
    const effectVal = this.effectVal({ initiator, target, game });
    target.stats.down('hp', effectVal);

    const targets = this.getTargets();
    targets.forEach((target, index) => {
      this.runAoe(initiator, target, game, index);
    });
  }

  runAoe(initiator: Player, target: Player, game: GameService, index: number) {
    const multiplier = 1 - (index + 1) * 0.1; // -10% каждой следующей цели
    const effectVal = this.effectVal({ initiator, target, game });
    const hit = floatNumber(effectVal * multiplier);

    target.stats.down('hp', hit);

    this.status.expArr.push({
      initiator,
      target,
      val: hit,
      hp: target.stats.val('hp'),
    });
  }

  aoeEffectVal({ initiator, target, game } = this.params): number {
    const effect = this.getEffectVal({ initiator, target, game });
    return this.modifyEffect(effect, { initiator, target, game });
  }
}

export default new ChainLightning();
