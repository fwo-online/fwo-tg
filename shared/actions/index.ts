import type { CharacterClass } from '@/character';

export type Magic = {
  name: string;
  description: string;
  displayName: string;
  effect: string[];
  cost: number;
  costType: CostType;
  lvl: number;
  orderType: OrderType;
  effectType?: EffectType;
};

export type Skill = {
  name: string;
  displayName: string;
  description: string;
  classList: Partial<Record<CharacterClass, number>>;
  bonusCost: number[];
  orderType: OrderType;
  cost: number;
  costType: CostType;
};

export enum CostType {
  MP = 'mp',
  EN = 'en',
}

export enum OrderType {
  All = 'all',
  Any = 'any',
  Enemy = 'enemy',
  Self = 'self',
  Team = 'team',
  TeamExceptSelf = 'teamExceptSelf',
}

export enum EffectType {
  Acid = 'acid',
  Fire = 'fire',
  Lighting = 'lighting',
  Frost = 'frost',
  Physical = 'physical',
  Clear = 'clear',
}

export type ActionType =
  | 'magic'
  | 'dmg-magic'
  | 'dmg-magic-long'
  | 'aoe-dmg-magic'
  | 'magic-long'
  | 'skill'
  | 'phys'
  | 'heal-magic'
  | 'heal'
  | 'protect'
  | 'dodge'
  | 'passive'
  | 'regeneration';

export type ReservedFailReason =
  | 'NO_INITIATOR'
  | 'NO_TARGET'
  | 'NO_MANA'
  | 'NO_ENERGY'
  | 'CHANCE_FAIL'
  | 'GOD_FAIL'
  | 'HEAL_FAIL'
  | 'SKILL_FAIL'
  | 'PHYS_FAIL'
  | 'NO_WEAPON';
