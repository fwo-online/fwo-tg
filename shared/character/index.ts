import type { CharacterAttributes } from './characterAttributesSchema';
import { CharacterClass } from './characterClassSchema';
import type { Attributes } from '@/shared/attributes';

export const characterAttributesByClass: Record<CharacterClass, CharacterAttributes> = {
  [CharacterClass.Warrior]: {
    str: 10,
    dex: 8,
    int: 3,
    wis: 3,
    con: 6,
  },
  [CharacterClass.Archer]: {
    str: 3,
    dex: 10,
    int: 8,
    wis: 3,
    con: 6,
  },
  [CharacterClass.Mage]: {
    str: 3,
    dex: 3,
    int: 8,
    wis: 10,
    con: 6,
  },
  [CharacterClass.Priest]: {
    str: 3,
    dex: 3,
    int: 10,
    wis: 8,
    con: 6,
  },
};

export type CharacterDynamicAttributes = Attributes & {
  maxTarget: number;
  spellLength: number;
};

export * from './characterAttributesSchema';
export * from './characterClassSchema';
export * from './characterSchema';
export * from './createCharacterSchema';
export * from './characterPublic';
