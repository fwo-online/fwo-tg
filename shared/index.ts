export * from './character';
export * from './inventory';
export * from './item';
export * from './itemSet';
export * from './shared';
export * from './actions';
export * from './game';
export * from './clan';
