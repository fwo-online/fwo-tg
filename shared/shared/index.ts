export * from './id';
export * from './modifiers';
export * from './player';
export * from './orderSchema';
export * from './message';
export * from './attributes';
export * from './action';
