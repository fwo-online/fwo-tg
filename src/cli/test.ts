import { $ } from 'bun';

async function test() {
  try {
    await $`docker-compose up -d --build db`;

    await $`bun test --preload ./test.setup.ts`;

    await $`docker-compose stop db`;
    process.exit(0);
  } catch {
    process.exit(1);
  }
}

void test();
