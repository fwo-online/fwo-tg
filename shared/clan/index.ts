import { nameSchema } from '@/shared';
import * as v from 'valibot';

export const reservedClanName = '__clan';

export type Clan = {
  id: string;
  name: string;
  gold: number;
  lvl: number;
  owner: string;
  players: string[];
  requests: string[];
  hasEmptySlot: boolean;
  maxPlayers: number;
};

export type ClanPublic = Pick<Clan, 'id' | 'name' | 'players' | 'owner'>;

export const createClanSchema = v.object({
  name: nameSchema,
});

export const clanLvlCost = [100, 250, 750, 1500];
export const clanAcceptCostPerLvl = 50;
