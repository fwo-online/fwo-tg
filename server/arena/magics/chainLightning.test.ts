import { afterEach, beforeEach, describe, expect, it } from 'bun:test';
import { CharacterClass } from '@fwo/shared';
import { times } from 'lodash';
import type GameService from '@/arena/GameService';
import TestUtils from '@/utils/testUtils';
import chainLightning from './chainLightning';

// npm t server/arena/magics/chainLightning.test.ts

describe('chainLightning', () => {
  let game: GameService;

  beforeEach(async () => {
    game = await TestUtils.createGame([
      { prof: CharacterClass.Mage, magics: { chainLightning: 3 } },
      ...times(10, () => ({})),
    ]);
  });

  beforeEach(() => {
    TestUtils.mockRandom(0.15);
  });

  afterEach(() => {
    TestUtils.restoreRandom();
  });

  it('should hit 5 targets', () => {
    game.players.players[0].proc = 1;

    chainLightning.cast(game.players.players[0], game.players.players[1], game);

    expect(game.players.players.map((player) => player.stats.val('hp'))).toMatchSnapshot();
    expect(game.players.players[0].stats.val('exp')).toMatchSnapshot();
    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
  });

  it('should hit 4 targets', () => {
    game.players.players[0].magics = { chainLightning: 2 };
    game.players.players[0].proc = 1;

    chainLightning.cast(game.players.players[0], game.players.players[1], game);

    expect(game.players.players.map((player) => player.stats.val('hp'))).toMatchSnapshot();
    expect(game.players.players[0].stats.val('exp')).toMatchSnapshot();
    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
  });

  it('should hit 3 targets', () => {
    game.players.players[0].magics = { chainLightning: 1 };
    game.players.players[0].proc = 1;

    chainLightning.cast(game.players.players[0], game.players.players[1], game);

    expect(game.players.players.map((player) => player.stats.val('hp'))).toMatchSnapshot();
    expect(game.players.players[0].stats.val('exp')).toMatchSnapshot();
    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
  });

  it('should hit 3 targets without clan', () => {
    game.players.players[0].magics = { chainLightning: 1 };
    game.players.players[0].proc = 1;

    chainLightning.cast(game.players.players[0], game.players.players[8], game);

    expect(game.players.players.map((player) => player.stats.val('hp'))).toMatchSnapshot();
    expect(game.players.players[0].stats.val('exp')).toMatchSnapshot();
    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
  });
});
