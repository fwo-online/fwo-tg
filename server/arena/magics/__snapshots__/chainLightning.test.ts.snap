// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chainLightning should hit 3 targets 1`] = `
[
  5.89,
  5.63,
  6.11,
  8,
  8,
  8,
  8,
  8,
  8,
  8,
  8,
]
`;

exports[`chainLightning should hit 3 targets 2`] = `63`;

exports[`chainLightning should hit 3 targets 3`] = `
"*architecto* сотворил _Цепь молний_ на *aspernatur* нанеся 6.36 урона
\\[ aspernatur ⚡ 💔-2.36/5.63 📖31
architecto ⚡ 💔-2.11/5.89 📖17
quia ⚡ 💔-1.89/6.11 📖15 ]"
`;

exports[`chainLightning should hit 3 targets without clan 1`] = `
[
  5.89,
  6.13,
  8,
  8,
  8,
  8,
  8,
  8,
  5.66,
  8,
  8,
]
`;

exports[`chainLightning should hit 3 targets without clan 2`] = `63`;

exports[`chainLightning should hit 3 targets without clan 3`] = `
"*qui* сотворил _Цепь молний_ на *ex* нанеся 6.32 урона
\\[ ex ⚡ 💔-2.34/5.66 📖31
qui ⚡ 💔-2.11/5.89 📖17
doloremque ⚡ 💔-1.87/6.13 📖15 ]"
`;

exports[`chainLightning should hit 4 targets 1`] = `
[
  4.98,
  4.64,
  5.31,
  5.62,
  8,
  8,
  8,
  8,
  8,
  8,
  8,
]
`;

exports[`chainLightning should hit 4 targets 2`] = `104`;

exports[`chainLightning should hit 4 targets 3`] = `
"*explicabo* сотворил _Цепь молний_ на *et* нанеся 11.49 урона
\\[ et ⚡ 💔-3.4/4.64 📖39
explicabo ⚡ 💔-3.02/4.98 📖24
eum ⚡ 💔-2.69/5.31 📖22
veritatis ⚡ 💔-2.38/5.62 📖19 ]"
`;

exports[`chainLightning should hit 5 targets 1`] = `
[
  4.06,
  3.59,
  4.47,
  4.9,
  5.37,
  8,
  8,
  8,
  8,
  8,
  8,
]
`;

exports[`chainLightning should hit 5 targets 2`] = `153`;

exports[`chainLightning should hit 5 targets 3`] = `
"*asperiores* сотворил _Цепь молний_ на *magni* нанеся 17.58 урона
\\[ magni ⚡ 💔-4.38/3.59 📖47
asperiores ⚡ 💔-3.94/4.06 📖32
labore ⚡ 💔-3.53/4.47 📖28
iure ⚡ 💔-3.1/4.9 📖25
recusandae ⚡ 💔-2.63/5.37 📖21 ]"
`;
